src/pybind/mgr/dashboard/controllers/task.py

```diff
@@ -13,3 +14,15 @@ def list(self, name=None):
             'executing_tasks': executing_t,
             'finished_tasks': finished_t
         }
+
+@ApiController('/progress')
+class TaskProgress(BaseController):
+    @Endpoint()
+    def events(self, name=None):
+        mon_status = CephService.send_command(
+            'mon', "status")
```

Comment:
I tried fetching data from progress module initially but due to lack of proper docstring for the methods, I didn't understand which arguments I had to pass. To initialize progress module's Module class that inherits from MgrModule, I was/am unable to find out about those mandatory arguments for initiliazing MgrModule. 
From progress module:
    def __init__(self, *args, **kwargs):
        super(Module, self).__init__(*args, **kwargs)

And init of MgrModule looks like:

    def __init__(self, module_name, py_modules_ptr, this_ptr):

where I cannot find any info on arguments 'py_modules_ptr' and 'this_ptr'. Also I am unable to find any kind of helper documentation for progress module.

Reply:
Yes, you're right. I was checking if the Mgr API would provide an API for this, but I only found it for clearing/updating/completing progress, but not for retrieving it.  There's an non-Python exposed (C++) method for retrieving it (`ActivePyModules::get_progress_events()`), but it would require a Python wrapping (similar to `BaseMgrModule::ceph_complete_progress_event()`).

My concern is that by calling the monitor we're currently doing the following:
``` 
dashboard <--> Py mgr <--> C++ mgr <--> Monitor cluster <--> C++ mgr
```
When (if I'm not wrong) the following would be enough:
``` 
dashboard <--> Py mgr <--> C++ mgr
```
I remember @jcsp said that we shouldn't 'abuse' sending commands to monitor, as it might have performance impact on the cluster. In this case we are polling every 5 secs / user / dashboard tab open to the monitor.

Alternative we might use `send_command` to the mgr with the `progress json` command instead?

Reply:
Upon looking and testing further, I found that the commands sent via CephService's send_command (which call mgr module's send_command) are all sent to the monitor. I tried changing the service name to "mgr" but it threw me an error saying that it's not recognized. I found out that there's a function called json_command in ceph_argparse module that uses its send_command function which seems to accept commands to mgr. they look like so:

def json_command(cluster, target=('mon', ''), prefix=None, argdict=None,
inbuf=b'', timeout=0, verbose=False):
"""
Serialize a command and up a JSON command and send it with send_command() above.

def send_command(cluster, target=('mon', ''), cmd=None, inbuf=b'', timeout=0,
verbose=False):
"""
Send a command to a daemon using librados's
mon_command, osd_command, mgr_command, or pg_command. Any bulk input data
comes in inbuf."""

and in send_command, there's :
elif target[0] == 'mgr':
ret, outbuf, outs = run_in_thread(
cluster.mgr_command, cmd, inbuf, timeout=timeout)

The problem is that I am not sure what cluster I should pass. Am I in the right path? Thanks.

Reply:
The mgr module already provides a rados handler (`self.rados`) and you may use [`rados.mgr_command()`](http://docs.ceph.com/docs/master/rados/api/python/#rados.Rados.mgr_command)?